lib/jekyll/filters.rb

```diff
@@ -13,10 +13,12 @@ module Filters
     # input - The Markdown String to convert.
     #
     # Returns the HTML formatted String.
-    def markdownify(input)
+    def markdownify(input, mode = nil)
       site = @context.registers[:site]
       converter = site.find_converter_instance(Jekyll::Converters::Markdown)
-      converter.convert(input.to_s)
+      output = converter.convert(input.to_s)
+      output = output.gsub(%r!<p>(.*?)<\/p>\n!im, '\1') if mode.to_s == "inline"
```

Comment:
This is a possibility I overlooked, but you're right there is a problem here.
Removing only the first and last one `p` tag would not really make it better, as it'd potentially leave unopened/unclosed tags in the middle...